### DISCUSSION TITLE
did anybody bump their toe on this: deploying with Kamal fails

### DISCUSSION THREAD

**Discussion by wdiechmann** [rubymonolith/superform#37] [Q&A] [answered]
TL;DR; anybody having troubles building containers with superform in your Gemfile?

my pronouns are

```
kamal 1.3.1
ruby 3.2.2 (2023-03-30 revision e51014f9c0) [arm64-darwin22]
Rails 7.1.3.2
```

😎

\- and I'm on the 'ozempic for Rails views' since 2+ yrs, probably even longer 'cause I stepped into the "game" when Phlex's handling of form_with was sub par forcing me to roll my own.

Cue `superform` - as I wanted to move closer to 'std phlex' I fought to wrap around the `form_with` and huzzah: superform to the rescue.

I've been deploying to ARM for 5-6 months now using Kamal and enjoying the experience - that stint sadly did end abruptly yesterday after hours when I tried to deploy a commit with a 

```
√ src % tail -3 Gemfile
gem "pagy", "~> 8.2"

gem "superform", "~> 0.4.6"
```

One minute my output looked like

```
  INFO container not ready (unhealthy), retrying in 3s (attempt 3/7)...
  INFO [50e32023] Running docker container ls --all --filter name=^healthcheck-mortimer-3c930777b82809446a784c5b93162c2043c8f3b3$ --quiet | xargs docker inspect --format '{{if .State.Health}}{{.State.Health.Status}}{{else}}{{.State.Status}}{{end}}' on 135.181.202.106
  INFO [50e32023] Finished in 0.120 seconds with exit status 0 (successful).
  INFO Container is healthy!
  INFO [e9e06f40] Running docker container ls --all --filter name=^healthcheck-mortimer-3c930777b82809446a784c5b93162c2043c8f3b3$ --quiet | xargs docker stop on 135.181.202.106
  INFO [e9e06f40] Finished in 0.948 seconds with exit status 0 (successful).
  INFO [0c060409] Running docker container ls --all --filter name=^healthcheck-mortimer-3c930777b82809446a784c5b93162c2043c8f3b3$ --quiet | xargs docker container rm on 135.181.202.106
  INFO [0c060409] Finished in 0.196 seconds with exit status 0 (successful).
Detect stale containers...
  INFO [b127fe90] Running docker ps --filter label=service=mortimer --filter label=role=web --format "{{.Names}}" | while read line; do echo ${line#mortimer-web-}; done on 135.181.202.106
  INFO [b127fe90] Finished in 0.576 seconds with exit status 0 (successful).
  INFO [df6cc905] Running docker ps --filter label=service=mortimer --filter label=role=worker --format "{{.Names}}" | while read line; do echo ${line#mortimer-worker-}; done on 135.181.202.106
  INFO [df6cc905] Finished in 0.150 seconds with exit status 0 (successful).
Start container with version 3c930777b82809446a784c5b93162c2043c8f3b3 using a 7s readiness delay (or reboot if already running)...
  INFO [02be79ff] Running docker tag wdiechmann/mortimer:3c930777b82809446a784c5b93162c2043c8f3b3 wdiechmann/mortimer:latest on 135.181.202.106
  INFO [02be79ff] Finished in 0.160 seconds with exit status 0 (successful).
  INFO [9e2f8a2e] Running docker container ls --all --filter name=^mortimer-web-3c930777b82809446a784c5b93162c2043c8f3b3$ --quiet on 135.181.202.106
  INFO [9e2f8a2e] Finished in 0.152 seconds with exit status 0 (successful).
√ src % 
```

Next thing I saw was

```
  INFO [cdf7e33b] Finished in 0.109 seconds with exit status 0 (successful).
  INFO container not ready (unhealthy), retrying in 7s (attempt 7/7)...
  INFO [649543d0] Running docker container ls --all --filter name=^healthcheck-mortimer-32829689c0b101ca1fbb044f28cb894563eb6f07$ --quiet | xargs docker inspect --format '{{if .State.Health}}{{.State.Health.Status}}{{else}}{{.State.Status}}{{end}}' on 135.181.202.106
  INFO [649543d0] Finished in 0.105 seconds with exit status 0 (successful).
  INFO [b42be921] Running docker container ls --all --filter name=^healthcheck-mortimer-32829689c0b101ca1fbb044f28cb894563eb6f07$ --quiet | xargs docker logs --tail 50 2>&1 on 135.181.202.106
  INFO [b42be921] Finished in 0.216 seconds with exit status 0 (successful).
 ERROR from /usr/local/bundle/ruby/3.2.0/gems/zeitwerk-2.6.13/lib/zeitwerk/loader/eager_load.rb:10:in `synchronize'
        from /usr/local/bundle/ruby/3.2.0/gems/zeitwerk-2.6.13/lib/zeitwerk/loader/eager_load.rb:10:in `eager_load'
        from /usr/local/bundle/ruby/3.2.0/gems/zeitwerk-2.6.13/lib/zeitwerk/loader.rb:379:in `block in eager_load_all'
        from /usr/local/bundle/ruby/3.2.0/gems/zeitwerk-2.6.13/lib/zeitwerk/loader.rb:377:in `each'
        from /usr/local/bundle/ruby/3.2.0/gems/zeitwerk-2.6.13/lib/zeitwerk/loader.rb:377:in `eager_load_all'
        from /usr/local/bundle/ruby/3.2.0/gems/railties-7.1.3.2/lib/rails/application/finisher.rb:80:in `block in <module:Finisher>'
        from /usr/local/bundle/ruby/3.2.0/gems/railties-7.1.3.2/lib/rails/initializable.rb:32:in `instance_exec'
        from /usr/local/bundle/ruby/3.2.0/gems/railties-7.1.3.2/lib/rails/initializable.rb:32:in `run'
        from /usr/local/bundle/ruby/3.2.0/gems/railties-7.1.3.2/lib/rails/initializable.rb:61:in `block in run_initializers'
        from /usr/local/lib/ruby/3.2.0/tsort.rb:228:in `block in tsort_each'
        from /usr/local/lib/ruby/3.2.0/tsort.rb:350:in `block (2 levels) in each_strongly_connected_component'
        from /usr/local/lib/ruby/3.2.0/tsort.rb:431:in `each_strongly_connected_component_from'
        from /usr/local/lib/ruby/3.2.0/tsort.rb:349:in `block in each_strongly_connected_component'
        from /usr/local/lib/ruby/3.2.0/tsort.rb:347:in `each'
        from /usr/local/lib/ruby/3.2.0/tsort.rb:347:in `call'
        from /usr/local/lib/ruby/3.2.0/tsort.rb:347:in `each_strongly_connected_component'
        from /usr/local/lib/ruby/3.2.0/tsort.rb:226:in `tsort_each'
        from /usr/local/lib/ruby/3.2.0/tsort.rb:205:in `tsort_each'
        from /usr/local/bundle/ruby/3.2.0/gems/railties-7.1.3.2/lib/rails/initializable.rb:60:in `run_initializers'
        from /usr/local/bundle/ruby/3.2.0/gems/railties-7.1.3.2/lib/rails/application.rb:426:in `initialize!'
        from /rails/config/environment.rb:16:in `<main>'
        from config.ru:3:in `require_relative'
        from config.ru:3:in `block in <main>'
        from /usr/local/bundle/ruby/3.2.0/gems/rack-3.0.10/lib/rack/builder.rb:103:in `eval'
        from /usr/local/bundle/ruby/3.2.0/gems/rack-3.0.10/lib/rack/builder.rb:103:in `new_from_string'
        from /usr/local/bundle/ruby/3.2.0/gems/rack-3.0.10/lib/rack/builder.rb:94:in `load_file'
        from /usr/local/bundle/ruby/3.2.0/gems/rack-3.0.10/lib/rack/builder.rb:64:in `parse_file'
        from /usr/local/bundle/ruby/3.2.0/gems/rackup-2.1.0/lib/rackup/server.rb:354:in `build_app_and_options_from_config'
        from /usr/local/bundle/ruby/3.2.0/gems/rackup-2.1.0/lib/rackup/server.rb:263:in `app'
        from /usr/local/bundle/ruby/3.2.0/gems/rackup-2.1.0/lib/rackup/server.rb:424:in `wrapped_app'
        from /usr/local/bundle/ruby/3.2.0/gems/rackup-2.1.0/lib/rackup/server.rb:326:in `block in start'
        from /usr/local/bundle/ruby/3.2.0/gems/rackup-2.1.0/lib/rackup/server.rb:382:in `handle_profiling'
        from /usr/local/bundle/ruby/3.2.0/gems/rackup-2.1.0/lib/rackup/server.rb:325:in `start'
        from /usr/local/bundle/ruby/3.2.0/gems/railties-7.1.3.2/lib/rails/commands/server/server_command.rb:38:in `start'
        from /usr/local/bundle/ruby/3.2.0/gems/railties-7.1.3.2/lib/rails/commands/server/server_command.rb:145:in `block in perform'
        from <internal:kernel>:90:in `tap'
        from /usr/local/bundle/ruby/3.2.0/gems/railties-7.1.3.2/lib/rails/commands/server/server_command.rb:136:in `perform'
        from /usr/local/bundle/ruby/3.2.0/gems/thor-1.3.1/lib/thor/command.rb:28:in `run'
        from /usr/local/bundle/ruby/3.2.0/gems/thor-1.3.1/lib/thor/invocation.rb:127:in `invoke_command'
        from /usr/local/bundle/ruby/3.2.0/gems/railties-7.1.3.2/lib/rails/command/base.rb:178:in `invoke_command'
        from /usr/local/bundle/ruby/3.2.0/gems/thor-1.3.1/lib/thor.rb:527:in `dispatch'
        from /usr/local/bundle/ruby/3.2.0/gems/railties-7.1.3.2/lib/rails/command/base.rb:73:in `perform'
        from /usr/local/bundle/ruby/3.2.0/gems/railties-7.1.3.2/lib/rails/command.rb:71:in `block in invoke'
        from /usr/local/bundle/ruby/3.2.0/gems/railties-7.1.3.2/lib/rails/command.rb:149:in `with_argv'
        from /usr/local/bundle/ruby/3.2.0/gems/railties-7.1.3.2/lib/rails/command.rb:69:in `invoke'
        from /usr/local/bundle/ruby/3.2.0/gems/railties-7.1.3.2/lib/rails/commands.rb:18:in `<main>'
        from <internal:/usr/local/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:37:in `require'
        from <internal:/usr/local/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:37:in `require'
        from /usr/local/bundle/ruby/3.2.0/gems/bootsnap-1.18.3/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
        from /rails/bin/rails:4:in `<main>'
  ERROR {
  "Status": "unhealthy",
  "FailingStreak": 3,
  "Log": [
    {
      "Start": "2024-04-23T06:09:46.043651818Z",
      "End": "2024-04-23T06:09:46.108516557Z",
      "ExitCode": 1,
      "Output": "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n                                 Dload  Upload   Total   Spent    Left  Speed\n\r  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\r  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\ncurl: (7) Failed to connect to localhost port 3000 after 5 ms: Couldn't connect to server\n"
    },
    {
      "Start": "2024-04-23T06:09:47.109125221Z",
      "End": "2024-04-23T06:09:47.179874788Z",
      "ExitCode": 1,
      "Output": "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n                                 Dload  Upload   Total   Spent    Left  Speed\n\r  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\r  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\ncurl: (7) Failed to connect to localhost port 3000 after 6 ms: Couldn't connect to server\n"
    },
    {
      "Start": "2024-04-23T06:09:48.180361773Z",
      "End": "2024-04-23T06:09:48.267953378Z",
      "ExitCode": 1,
      "Output": "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n                                 Dload  Upload   Total   Spent    Left  Speed\n\r  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\r  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\ncurl: (7) Failed to connect to localhost port 3000 after 14 ms: Couldn't connect to server\n"
    }
  ]
}
  INFO [a3f80f55] Running docker container ls --all --filter name=^healthcheck-mortimer-32829689c0b101ca1fbb044f28cb894563eb6f07$ --quiet | xargs docker stop on 135.181.202.106
  INFO [a3f80f55] Finished in 0.162 seconds with exit status 0 (successful).
  INFO [e335531c] Running docker container ls --all --filter name=^healthcheck-mortimer-32829689c0b101ca1fbb044f28cb894563eb6f07$ --quiet | xargs docker container rm on 135.181.202.106
  INFO [e335531c] Finished in 0.183 seconds with exit status 0 (successful).
Releasing the deploy lock...
  Finished all in 138.3 seconds
  ERROR (Kamal::Cli::Healthcheck::Poller::HealthcheckError): Exception while executing on host 135.181.202.106: container not ready (unhealthy)
?1 src % 
```

**Comment by wdiechmann** [comment, marked as answer]
I'll close this discussion with a weird taste - went through the entire process of settting up a fresh VM, forging a new rails app with the bare minimum of gems - and to my great disappointment it deploys just fine with superform 😅

so it's back to the workbench looking for what possible other gem could be cross with me 🥹

**Reply by wdiechmann** [reply]
not really a reply nor an answer - but issue solved (but I can't keep starting all over everytime some deployment goes south now, can I)?
